File: Audio/Audio/AdpcmCodec.swift
import Foundation

/// IMA-ADPCM 编解码。与心声服务器/硬件端逐字节兼容。
/// 251 字节包：魔数"ADPC" + 版本 + 采样数(480) + 首采样 + 索引 + 保留 + 240字节数据。
final class AdpcmCodec {
    static let packetBytes = 251
    static let pcmFrameBytes = 960
    static let samplesPerFrame = 480
    private static let magic: [UInt8] = [65, 68, 80, 67] // "ADPC"

    private static let indexTable: [Int] = [-1,-1,-1,-1,2,4,6,8,-1,-1,-1,-1,2,4,6,8]
    private static let stepTable: [Int] = [
        7,8,9,10,11,12,13,14,16,17,19,21,23,25,28,31,34,37,41,45,50,55,60,66,73,80,88,97,
        107,118,130,143,157,173,190,209,230,253,279,307,337,371,408,449,494,544,598,658,
        724,796,876,963,1060,1166,1282,1411,1552,1707,1878,2066,2272,2499,2749,3024,3327,
        3660,4026,4428,4871,5358,5894,6484,7132,7845,8630,9493,10442,11487,12635,13899,
        15289,16818,18500,20350,22385,24623,27086,29794,32767
    ]

    private var carryIndex = 0
    func resetStreamState() { carryIndex = 0 }

    static func isAdpcmPacket(_ d: [UInt8]) -> Bool {
        d.count >= 8 && Array(d[0..<4]) == magic
    }

    /// 480 个 Int16 (960字节小端) -> 251 字节包
    func encodeFrame(_ pcmLe: [UInt8]) -> [UInt8]? {
        guard pcmLe.count >= AdpcmCodec.pcmFrameBytes else { return nil }
        var samples = [Int16](repeating: 0, count: AdpcmCodec.samplesPerFrame)
        for i in 0..<AdpcmCodec.samplesPerFrame {
            let lo = Int(pcmLe[i*2]); let hi = Int(pcmLe[i*2+1])
            samples[i] = Int16(bitPattern: UInt16(lo | (hi << 8)))
        }
        var out = [UInt8](repeating: 0, count: AdpcmCodec.packetBytes)
        out[0] = 65; out[1] = 68; out[2] = 80; out[3] = 67
        out[4] = 1; out[5] = 0xE0; out[6] = 0x01
        encodeImaBlock(samples, &out, 7)
        return out
    }

    static func decodeToPcm(_ data: [UInt8]) -> [UInt8]? {
        guard isAdpcmPacket(data), data.count >= packetBytes else { return nil }
        let samples = Int(data[5]) | (Int(data[6]) << 8)
        guard samples == samplesPerFrame else { return nil }
        let pcm = decodeImaBlock(data, 7)
        var out = [UInt8](repeating: 0, count: pcmFrameBytes)
        for i in 0..<samplesPerFrame {
            let v = UInt16(bitPattern: pcm[i])
            out[i*2] = UInt8(v & 0xFF); out[i*2+1] = UInt8((v >> 8) & 0xFF)
        }
        return out
    }

    private func encodeImaBlock(_ samples: [Int16], _ out: inout [UInt8], _ offset: Int) {
        let s = Int(samples[0])
        var index = carryIndex
        out[offset] = UInt8(s & 0xFF)
        out[offset+1] = UInt8((s >> 8) & 0xFF)
        out[offset+2] = UInt8(index & 0xFF)
        out[offset+3] = 0
        var step = AdpcmCodec.stepTable[index]
        var bytePos = offset + 4
        var buffer = 0, bufferBits = 0
        var prev = s
        for n in 1..<AdpcmCodec.samplesPerFrame {
            var diff = Int(samples[n]) - prev
            var sign = 0
            if diff < 0 { sign = 8; diff = -diff }
            var code = 0
            var vpdiff = step >> 3
            if diff >= step { code = 4; diff -= step; vpdiff += step }
            let half = step >> 1
            if diff >= half { code |= 2; diff -= half; vpdiff += half }
            let quarter = step >> 2
            if diff >= quarter { code |= 1; vpdiff += quarter }
            prev = AdpcmCodec.clamp16(sign != 0 ? prev - vpdiff : prev + vpdiff)
            index += AdpcmCodec.indexTable[code | sign]
            if index < 0 { index = 0 }; if index > 88 { index = 88 }
            step = AdpcmCodec.stepTable[index]
            buffer |= (code | sign) << bufferBits
            bufferBits += 4
            if bufferBits >= 8 { out[bytePos] = UInt8(buffer & 0xFF); bufferBits -= 8; buffer >>= 8; bytePos += 1 }
        }
        if bufferBits > 0 { out[bytePos] = UInt8(buffer & 0xFF) }
        carryIndex = index
    }

    private static func decodeImaBlock(_ data: [UInt8], _ offset: Int) -> [Int16] {
        var pcm = [Int16](repeating: 0, count: samplesPerFrame)
        var valpred = Int(Int16(bitPattern: UInt16(Int(data[offset]) | (Int(data[offset+1]) << 8))))
        var index = Int(data[offset+2]); if index < 0 { index = 0 }; if index > 88 { index = 88 }
        pcm[0] = Int16(truncatingIfNeeded: valpred)
        var step = stepTable[index]
        var bytePos = offset + 4
        var buffer = 0, bufferBits = 0
        for n in 1..<samplesPerFrame {
            if bufferBits < 4 { buffer |= Int(data[bytePos]) << bufferBits; bufferBits += 8; bytePos += 1 }
            let code = buffer & 0x0F
            buffer >>= 4; bufferBits -= 4
            index += indexTable[code]; if index < 0 { index = 0 }; if index > 88 { index = 88 }
            var vpdiff = step >> 3
            if code & 4 != 0 { vpdiff += step }
            if code & 2 != 0 { vpdiff += step >> 1 }
            if code & 1 != 0 { vpdiff += step >> 2 }
            step = stepTable[index]
            valpred = clamp16(code & 8 != 0 ? valpred - vpdiff : valpred + vpdiff)
            pcm[n] = Int16(truncatingIfNeeded: valpred)
        }
        return pcm
    }

    private static func clamp16(_ v: Int) -> Int { v > 32767 ? 32767 : (v < -32768 ? -32768 : v) }
}

File: android/app/src/main/java/com/coco/audio/audio/AdpcmCodec.kt
package com.coco.audio.audio

import java.nio.ByteBuffer
import java.nio.ByteOrder

/**
 * IMA-ADPCM 编解码。与心声服务器/硬件端逐字节兼容。
 * 每个数据包 251 字节：4字节魔数"ADPC" + 版本 + 采样数(480) + 首采样值 + 步长索引 + 保留 + 240字节数据。
 */
object AdpcmCodec {
    const val PACKET_BYTES = 251
    const val PCM_FRAME_BYTES = 960
    const val SAMPLES_PER_FRAME = 480
    private const val VERSION: Byte = 1
    private val MAGIC = byteArrayOf(65, 68, 80, 67) // "ADPC"

    private val INDEX_TABLE = intArrayOf(-1, -1, -1, -1, 2, 4, 6, 8, -1, -1, -1, -1, 2, 4, 6, 8)
    private val STEP_TABLE = intArrayOf(
        7, 8, 9, 10, 11, 12, 13, 14, 16, 17, 19, 21, 23, 25, 28, 31, 34, 37, 41, 45, 50, 55, 60, 66,
        73, 80, 88, 97, 107, 118, 130, 143, 157, 173, 190, 209, 230, 253, 279, 307, 337, 371, 408,
        449, 494, 544, 598, 658, 724, 796, 876, 963, 1060, 1166, 1282, 1411, 1552, 1707, 1878, 2066,
        2272, 2499, 2749, 3024, 3327, 3660, 4026, 4428, 4871, 5358, 5894, 6484, 7132, 7845, 8630,
        9493, 10442, 11487, 12635, 13899, 15289, 16818, 18500, 20350, 22385, 24623, 27086, 29794, 32767
    )

    private var carryIndex = 0

    fun resetStreamState() { carryIndex = 0 }

    fun isAdpcmPacket(data: ByteArray): Boolean =
        data.size >= 8 && data.copyOfRange(0, 4).contentEquals(MAGIC)

    fun encodeFrame(pcmLe: ByteArray): ByteArray? {
        if (pcmLe.size < PCM_FRAME_BYTES) return null
        val pcm = ByteBuffer.wrap(pcmLe).order(ByteOrder.LITTLE_ENDIAN).asShortBuffer()
        val samples = ShortArray(SAMPLES_PER_FRAME)
        pcm.get(samples, 0, SAMPLES_PER_FRAME)
        val out = ByteArray(PACKET_BYTES)
        System.arraycopy(MAGIC, 0, out, 0, 4)
        out[4] = VERSION
        out[5] = 0xE0.toByte()   // 480 low
        out[6] = 0x01            // 480 high
        encodeImaBlock(samples, out, 7)
        return out
    }

    fun decodeToPcm(data: ByteArray): ByteArray? {
        if (!isAdpcmPacket(data) || data.size < PACKET_BYTES) return null
        val samples = (data[5].toInt() and 0xFF) or ((data[6].toInt() and 0xFF) shl 8)
        if (samples != SAMPLES_PER_FRAME) return null
        val pcm = decodeImaBlock(data, 7)
        val buf = ByteBuffer.allocate(PCM_FRAME_BYTES).order(ByteOrder.LITTLE_ENDIAN)
        for (s in pcm) buf.putShort(s)
        return buf.array()
    }

    private fun encodeImaBlock(samples: ShortArray, out: ByteArray, offset: Int) {
        val s = samples[0]
        var index = carryIndex
        out[offset] = (s.toInt() and 0xFF).toByte()
        out[offset + 1] = ((s.toInt() shr 8) and 0xFF).toByte()
        out[offset + 2] = index.toByte()
        out[offset + 3] = 0
        var step = STEP_TABLE[index]
        var bytePos = offset + 4
        var buffer = 0
        var bufferBits = 0
        var prev = s.toInt()
        for (n in 1 until SAMPLES_PER_FRAME) {
            var diff = samples[n] - prev
            var sign = 0
            if (diff < 0) { sign = 8; diff = -diff }
            var code = 0
            var vpdiff = step shr 3
            if (diff >= step) { code = 4; diff -= step; vpdiff += step }
            val half = step shr 1
            if (diff >= half) { code = code or 2; diff -= half; vpdiff += half }
            val quarter = step shr 2
            if (diff >= quarter) { code = code or 1; vpdiff += quarter }
            prev = clamp16(if (sign != 0) prev - vpdiff else prev + vpdiff)
            index += INDEX_TABLE[code or sign]
            if (index < 0) index = 0
            if (index > 88) index = 88
            step = STEP_TABLE[index]
            buffer = buffer or ((code or sign) shl bufferBits)
            bufferBits += 4
            if (bufferBits >= 8) {
                out[bytePos] = (buffer and 0xFF).toByte()
                bufferBits -= 8
                buffer = buffer shr 8
                bytePos++
            }
        }
        if (bufferBits > 0) out[bytePos] = (buffer and 0xFF).toByte()
        carryIndex = index
    }

    private fun decodeImaBlock(data: ByteArray, offset: Int): ShortArray {
        val pcm = ShortArray(SAMPLES_PER_FRAME)
        var valpred = ((data[offset].toInt() and 0xFF) or ((data[offset + 1].toInt() and 0xFF) shl 8)).toShort().toInt()
        var index = data[offset + 2].toInt()
        if (index < 0) index = 0
        if (index > 88) index = 88
        pcm[0] = valpred.toShort()
        var step = STEP_TABLE[index]
        var bytePos = offset + 4
        var buffer = 0
        var bufferBits = 0
        for (n in 1 until SAMPLES_PER_FRAME) {
            if (bufferBits < 4) {
                buffer = buffer or ((data[bytePos].toInt() and 0xFF) shl bufferBits)
                bufferBits += 8
                bytePos++
            }
            val code = buffer and 0x0F
            buffer = buffer shr 4
            bufferBits -= 4
            index += INDEX_TABLE[code]
            if (index < 0) index = 0
            if (index > 88) index = 88
            var vpdiff = step shr 3
            if (code and 4 != 0) vpdiff += step
            if (code and 2 != 0) vpdiff += step shr 1
            if (code and 1 != 0) vpdiff += step shr 2
            step = STEP_TABLE[index]
            valpred = clamp16(if (code and 8 != 0) valpred - vpdiff else valpred + vpdiff)
            pcm[n] = valpred.toShort()
        }
        return pcm
    }

    private fun clamp16(v: Int): Int = when {
        v > 32767 -> 32767
        v < -32768 -> -32768
        else -> v
    }
}
